### PR TITLE
First step to get backend speaking right language to clients

### PR DESCRIPTION
**Pull request by andrewconner**
Please help me verify that this makes no changes to current behavior. Next step, I'll move the site to use the POST /add endpoint, which speaks JSON.

For Ryan, problem was that requests come from two sources: things like browser clients (as if you typed in the URL and hit enter) and API clients (how the site's code talks to the backend). Browsers speak HTML, never speak JSON (because they'll just display it as text, and users get confused), and can deal with redirects to HTML. API endpoints speak JSON always, and cannot deal with redirects to HTML pages. As a rule, `/site/` routes are always the latter, for technical and convention reasons.

We accidentally had a little mix-and-match, which means sometimes users might see JSON, or our API gets confused why it's getting back HTML and barfs. This doesn't really fix anything, but lets me do the client side on the /add endpoint, and hopefully makes the code a bit more clear‽

@leogrim @ryanpbrewster Please read over, thanks.

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/SlackController.scala [hunk 1]
```diff
@@ -64,34 +64,39 @@ class SlackController @Inject() (
...
+      action <- slackStateCommander.getSlackAction(SlackAuthState(state)).map(Future.successful).getOrElse(Future.failed(SlackAPIFailure.InvalidAuthState))
       slackAuth <- slackClient.processAuthorizationResponse(SlackAuthorizationCode(code), SlackController.REDIRECT_URI)
       slackIdentity <- slackClient.identifyUser(slackAuth.accessToken)
       result <- {
         slackCommander.registerAuthorization(request.userIdOpt, slackAuth, slackIdentity)
         processAuthorizedAction(request.userId, slackIdentity.teamId, slackIdentity.userId, action, slackAuth.incomingWebhook)
       }
-    } yield result
+    } yield {
+      result match {
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/SlackController.scala:75]: I am not certain this is correct.

Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/SlackController.scala:75]: What is the type of `result`?

Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/SlackController.scala:75]: We should be able to reconstruct where the user when they initiated the `AuthorizedAction`, and maybe we can redirect them back there with a url param encoding the error?

File: kifi-backend/modules/shoebox/conf/shoeboxService.routes [hunk 2]
```diff
@@ -555,11 +556,15 @@ GET           /site/organizations/:id/referralCode              @com.keepit.cont
...
+# ↓ Take out of /site/
 GET           /site/slack/add                                   @com.keepit.controllers.website.SlackController.addSlackTeam(slackTeamId: Option[String] ?= None)
 GET           /site/slack/add/organizations                     @com.keepit.controllers.website.SlackController.getOrganizationsToConnectToSlackTeam()
+# ↓ Take out of /site/
 GET           /site/organizations/create/slack                  @com.keepit.controllers.website.SlackController.createSlackTeam(slackTeamId: Option[String] ?= None)
+# ↓ Take out of /site/
 GET           /site/organizations/:id/slack/connect             @com.keepit.controllers.website.SlackController.connectSlackTeam(id: PublicId[Organization], slackTeamId: Option[String] ?= None)
 GET           /site/organizations/:id/slack/list                @com.keepit.controllers.website.SlackController.getOrgIntegrations(id: PublicId[Organization])
+# ↓ Take out of /site/ OR change to be be JSON endpoint
 GET           /site/organizations/:id/slack/sync/public         @com.keepit.controllers.website.SlackController.syncPublicChannels(id: PublicId[Organization])
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/conf/shoeboxService.routes:568]: I think all of these make sense inside of the `site.routes` file, right next to the existing `SlackAuthRouter` routes

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/SlackController.scala [hunk 3]
```diff
@@ -285,7 +302,33 @@ class SlackController @Inject() (
...
+          .map(handleAsBrowserRequest)
+      case Failure(_) => Future.successful(BadRequest("invalid_integration_id")) // Bad bad bad
     }
   }
+
+  // Can elegantly handle redirects (to HTML pages), *never* speaks JSON, should not be on /site/ routes
+  private def handleAsBrowserRequest[T](implicit request: Request[T]) = { (response: SlackResponse) =>
+    response match {
+      case SlackResponse.RedirectClient(url) => Redirect(url, SEE_OTHER)
+      case SlackResponse.ActionPerformed => Redirect(request.headers.get(REFERER).getOrElse("/")) // Bad bad bad
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/SlackController.scala:314]: Scary...

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/SlackController.scala [hunk 4]
```diff
@@ -285,7 +302,33 @@ class SlackController @Inject() (
...
+      case SlackResponse.Error(code) => Redirect("/") // Bad bad bad
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/SlackController.scala:315]: Can we have these things airbrake so we know if they're getting hit?

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/SlackController.scala (hunk 1 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/SlackController.scala:75]: lgtm

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/SlackController.scala [hunk 5]
```diff
@@ -285,7 +302,33 @@ class SlackController @Inject() (
...
+    }
+  }
+
+  // Always speaks JSON, should be on /site/ routes, cannot handle Redirects to HTML pages
+  private def handleAsAPIRequest[T](implicit request: Request[T]) = { (response: SlackResponse) =>
+    response match {
+      case SlackResponse.RedirectClient(url) => Ok(Json.obj("redirect" -> url))
+      case SlackResponse.ActionPerformed => Ok(Json.obj("success" -> true))
+      case SlackResponse.Error(code) => BadRequest(Json.obj("error" -> code))
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/SlackController.scala:324]: Low priority, but how would you feel about having `SlackResponse.Error` actually keep track of what the status is (similar to how we handle `OrganizationFail`)?

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/SlackController.scala [hunk 6]
```diff
@@ -285,7 +302,33 @@ class SlackController @Inject() (
...
+    }
+  }
+}
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/SlackController.scala:327]: :+1:

**Comment by leogrim** [comment]
Looks great, thanks for tidying this up!

**Comment by ryanpbrewster** [comment]
The OTF (one true flow) looks right. Error handling is hard, in these situation.

This seems strictly better than what we currently have, though.

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/SlackController.scala (hunk 5 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/SlackController.scala:324]: They are all `BadRequest` currently

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/SlackController.scala (hunk 4 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/SlackController.scala:315]: Currently, they might be hit, actually. Endpoints that could fail like that shouldn't be used in that way.

File: kifi-backend/modules/shoebox/conf/shoeboxService.routes (hunk 2 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/conf/shoeboxService.routes:568]: Yep, that's next step. Can't break these yet.

**Pull request merged**